### PR TITLE
mock: simplify implementation of FunctionalOptions

### PR DESCRIPTION
**Pull request by dolmen**
## Summary
Remove unnecessary use of [reflect](https://pkg.go.dev/reflect) in the implementation of [mock.FunctionalOptions()](https://pkg.go.dev/github.com/stretchr/testify/mock#FunctionalOptions).

## Changes
* Change the internal representation of `FunctionalOptionsArgument` to be a slice of arguments to match the variadic argument of `FunctionalOptions`.
* Remove unnecessary use of `reflect`.

## Motivation
* Simpler code
* Runtime speed

### CODE REVIEW

**Comment by dolmen** [comment]
@snirye Could you review this change as you are a `mock.FunctionalOptions` user?

**Comment by dolmen** [comment]
@nbaztec Could you review this change as you wrote the initial implementation?

**Review by brackendawson — DISMISSED** [review]
LGTM

I did some testing. No behavioural change but less uncecessary reflection.

**Review by andrewwillette — APPROVED** [review]

**Comment by dolmen** [comment]
Rebased (with conflict resolution) after merge of f1b5324b903e87676e98b4fa3a5e2fbe2c7dddc5 (#1433).

**Review by andrewwillette — APPROVED** [review]

**Comment by andrewwillette** [comment]
@dolmen can this be merged? I'm hoping to unblock [1578](https://github.com/stretchr/testify/issues/1578).

**Pull request merged**
